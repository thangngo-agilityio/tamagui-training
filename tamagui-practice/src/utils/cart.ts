// Types

import { ICartItem } from '@/types';

export const calculateTotalPrice = (cartItems: ICartItem[] = []) =>
  cartItems.reduce((acc, cartItem) => {
    const { product, quantity = 1 } = cartItem || {};
    const { price = 0 } = product || {};

    const itemPrice = price * quantity;

    return acc + itemPrice;
  }, 0);

export const calculateTotalQuantity = (cartItems: ICartItem[] = []) =>
  cartItems.reduce((acc, cartItem) => {
    const { quantity = 0 } = cartItem || {};

    return acc + quantity;
  }, 0);
