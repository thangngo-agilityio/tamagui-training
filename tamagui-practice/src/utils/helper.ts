export const formatAmountNumber = (value: string): string => {
  if (!value) {
    return value;
  }

  if (Number.isNaN(parseFloat(value))) {
    return '';
  }

  const dotIndex = value.indexOf('.');
  const newValue =
    dotIndex > -1
      ? value.slice(0, dotIndex).replaceAll(',', '')
      : value.replaceAll(',', '');

  const newValueFormat = newValue.replace(/\B(?=(\d{3})+(?!\d))/g, ',');

  return newValueFormat;
};
