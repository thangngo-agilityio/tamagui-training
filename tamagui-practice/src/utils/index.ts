export * from './images';
export * from './pagination';
export * from './paramUrl';
export * from './helper';
export * from './cart';
export * from './format';
export * from './validators';
