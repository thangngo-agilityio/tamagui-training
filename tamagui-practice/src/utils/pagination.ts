// Constants
import { DOTS } from '@/constants';

// Types
import { FormatPaginationParams, PaginationTableType } from '@/types';

export const formatNumberButton = (numberOfPage: number): string[] =>
  Array.from({ length: numberOfPage }).map((_, index: number): string =>
    (index + 1).toString(),
  );

export const formatPageArray = ({
  totalPage = 0,
  currentPage = 0,
}: PaginationTableType): string[] => {
  // If the number of pages is less than or equal to 5, display all page buttons
  if (totalPage <= 5) {
    return formatNumberButton(totalPage);
  }

  // If the current page is near the start (pages 1 to 3),
  // Display the first 3 pages and then DOTS and the last page
  if (currentPage <= 3) {
    return [...formatNumberButton(3), DOTS, totalPage.toString()];
  }

  // If the current page is near the end (within the last 3 pages),
  // display the first page, DOTS, and the last 3 pages
  if (currentPage >= totalPage - 2) {
    return ['1', DOTS, ...formatNumberButton(totalPage).splice(-3)];
  }

  // If the current page is in the middle,
  // Display the first page, DOTS, currentPage, DOTS, and the last page
  return ['1', DOTS, currentPage.toString(), DOTS, totalPage.toString()];
};

export const formatPagination = ({
  totalCount,
  limit,
  currentPage,
}: FormatPaginationParams): string[] => {
  const totalPage = Math.ceil(totalCount / limit);

  return formatPageArray({ totalPage, currentPage });
};
