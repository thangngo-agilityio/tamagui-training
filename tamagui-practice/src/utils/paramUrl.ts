export const getSearchParams = (
  searchParams: URLSearchParams,
): Record<string, string> => {
  const queryParam: Record<string, string> = {};

  searchParams.forEach((value, key) => {
    queryParam[key] = value;
  });

  return queryParam;
};

export const updateSearchParams = (
  searchParams: URLSearchParams,
  key: string,
  value: string,
): URLSearchParams => {
  const params = new URLSearchParams(searchParams);

  if (value) {
    params.set(key, value);
  } else {
    params.delete(key);
  }

  return params;
};

export const formatUrlWithQuery = (
  endpoint: string,
  query: Record<string, string | string[] | number | number[] | undefined>,
): string => {
  const params = new URLSearchParams(
    Object.entries(query)
      .filter(([, value]) => value !== undefined)
      .flatMap(([key, value]) =>
        Array.isArray(value)
          ? value.map((val) => [key, String(val)])
          : [[key, String(value)]],
      ),
  );

  return params.toString() ? `${endpoint}?${params.toString()}` : endpoint;
};
