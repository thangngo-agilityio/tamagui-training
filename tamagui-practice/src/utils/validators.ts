export const isEnableSubmitButton = (
  requiredFields: string[],
  dirtyFields: string[],
  errors: Record<string, unknown>,
): boolean => {
  const isMatchAllRequiredFields: boolean = requiredFields.every((field) =>
    dirtyFields.includes(field),
  );

  return isMatchAllRequiredFields && errors && !Object.keys(errors).length;
};
