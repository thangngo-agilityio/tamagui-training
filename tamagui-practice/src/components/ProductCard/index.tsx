import { Box, Flex, Heading, Text } from '@chakra-ui/react';
import Image from 'next/image';

// Icons
import { HeartIcon } from '@/icons';
import { formatAmountNumber } from '@/utils';
import Link from 'next/link';
import { ROUTER } from '@/constants';

type ProductCard = {
  id: string;
  image: string;
  title: string;
  price: number;
};

const ProductCard = ({ id, image, title, price }: ProductCard) => (
  <Link href={ROUTER.PRODUCT_DETAIL_PAGE(id)} scroll>
    <Flex
      flexDirection="column"
      borderRadius="sm"
      boxShadow="0 4px 50px -5px rgba(32, 32, 32, 10%)"
      pb={{ base: '8px', lg: '14px' }}
      alignItems="center"
    >
      <Flex
        w="90%"
        position="relative"
        justifyContent="center"
        mt="-90px"
        mb="18px"
        transition=".2s ease-in"
        _hover={{
          w: '100%',
        }}
      >
        <Box
          as="div"
          w="100%"
          h={{ base: '94px', lg: '180px' }}
          borderRadius="sm"
          overflow="hidden"
        >
          <Image
            width={276}
            height={180}
            src={image}
            alt={title}
            style={{
              borderRadius: 'sm',
              backgroundRepeat: 'no-repeat',
              width: '100%',
              height: '100%',
              objectFit: 'cover',
            }}
          />
        </Box>
        <Flex
          w={{ base: '30px', lg: '45px' }}
          h={{ base: '30px', lg: '45px' }}
          position="absolute"
          justifyContent="center"
          alignItems="center"
          borderRadius="full"
          bgColor="background.100"
          top="10px"
          right={{ base: '15px', lg: '30px' }}
        >
          <HeartIcon />
        </Flex>
      </Flex>
      <Flex w="100%" px={{ base: '8px', lg: '16px' }} flexDirection="column">
        <Heading variant="senary" size={{ base: 'sm', lg: 'xl' }} noOfLines={1}>
          {title}
        </Heading>
        <Text variant="productCard" size={{ base: 'textMd', lg: 'textLg' }}>
          N{formatAmountNumber(price.toString())}
        </Text>
      </Flex>
    </Flex>
  </Link>
);

export default ProductCard;
