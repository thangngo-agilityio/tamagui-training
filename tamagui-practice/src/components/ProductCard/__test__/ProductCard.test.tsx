// Unit tests for: ProductCard

import { formatAmountNumber } from '@/utils';
import { render } from '@testing-library/react';
import React from 'react';
import ProductCard from '../index';
import '@testing-library/jest-dom';

// Mocking the dependencies
jest.mock('next/image', () => ({ src, alt }: { src: string; alt: string }) => (
  <img src={src} alt={alt} />
));
jest.mock('@/icons', () => ({
  HeartIcon: () => <svg data-testid="heart-icon" />,
}));
jest.mock('@/utils', () => ({
  formatAmountNumber: jest.fn((price: string) => price),
}));
jest.mock(
  'next/link',
  () =>
    ({ children, href }: { children: React.ReactNode; href: string }) => (
      <a href={href}>{children}</a>
    ),
);
jest.mock('@/constants', () => ({
  ROUTER: { PRODUCT_DETAIL_PAGE: (id: string) => `/product/${id}` },
}));

describe('ProductCard() ProductCard method', () => {
  const product = {
    id: '1',
    image: '/path/to/image.jpg',
    title: 'Sample Product',
    price: 1000,
  };
  it('', () => {
    const { container } = render(<ProductCard {...product} />);

    expect(container).toMatchSnapshot();
  });

  // Happy Path Tests
  describe('Happy Path', () => {
    it('should render the product card with correct details', () => {
      // Act
      const { getByText, getByAltText, getByTestId } = render(
        <ProductCard {...product} />,
      );

      // Assert
      expect(getByAltText('Sample Product')).toHaveAttribute(
        'src',
        '/path/to/image.jpg',
      );
      expect(getByText('Sample Product')).toBeInTheDocument();
      expect(getByText('N1000')).toBeInTheDocument();
      expect(getByTestId('heart-icon')).toBeInTheDocument();
    });

    it('should link to the correct product detail page', () => {
      // Arrange
      const product = {
        id: '1',
        image: '/path/to/image.jpg',
        title: 'Sample Product',
        price: 1000,
      };

      // Act
      const { container } = render(<ProductCard {...product} />);
      const link = container.querySelector('a');

      // Assert
      expect(link).toHaveAttribute('href', '/product/1');
    });
  });

  // Edge Case Tests
  describe('Edge Cases', () => {
    it('should handle a product with a very long title gracefully', () => {
      // Arrange
      const product = {
        id: '1',
        image: '/path/to/image.jpg',
        title: 'This is a very long product title that should be truncated',
        price: 1000,
      };

      // Act
      const { getByText } = render(<ProductCard {...product} />);

      // Assert
      expect(
        getByText('This is a very long product title that should be truncated'),
      ).toBeInTheDocument();
    });

    it('should format the price correctly using formatAmountNumber', () => {
      // Arrange
      const product = {
        id: '1',
        image: '/path/to/image.jpg',
        title: 'Sample Product',
        price: 1000,
      };

      // Act
      render(<ProductCard {...product} />);

      // Assert
      expect(formatAmountNumber).toHaveBeenCalledWith('1000');
    });

    it('should render without crashing when image is missing', () => {
      // Arrange
      const product = {
        id: '1',
        image: '',
        title: 'Sample Product',
        price: 1000,
      };

      // Act
      const { getByAltText } = render(<ProductCard {...product} />);

      // Assert
      expect(getByAltText('Sample Product')).toHaveAttribute('src', '');
    });
  });
});

// End of unit tests for: ProductCard
