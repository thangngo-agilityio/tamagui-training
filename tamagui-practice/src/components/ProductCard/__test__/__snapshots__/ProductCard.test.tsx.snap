// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard() ProductCard method  1`] = `
<div>
  <a
    href="/product/1"
  >
    <div
      class="css-6rwikn"
    >
      <div
        class="css-12nnctg"
      >
        <div
          class="css-1vani7c"
        >
          <img
            alt="Sample Product"
            src="/path/to/image.jpg"
          />
        </div>
        <div
          class="css-fv60i0"
        >
          <svg
            data-testid="heart-icon"
          />
        </div>
      </div>
      <div
        class="css-p4lymy"
      >
        <h2
          class="chakra-heading css-zvlevn"
        >
          Sample Product
        </h2>
        <p
          class="chakra-text css-0"
        >
          N
          1000
        </p>
      </div>
    </div>
  </a>
</div>
`;
